internal/bufferedwrites/upload_handler.go

```diff
@@ -154,3 +155,39 @@ func (uh *UploadHandler) SignalUploadFailure() chan error {
 func (uh *UploadHandler) AwaitBlocksUpload() {
 	uh.wg.Wait()
 }
+
+func (uh *UploadHandler) Destroy() {
+	// Waiting for upload routine to move all blocks to freeChannel.
+	timedOut := waitTimeout(&uh.wg, 10*time.Second)
+
+	// TimedOut means there are some blocks which are still in uploadChannel,
+	// either because upload is stuck in uploading a chunk or the upload
+	// go-routine crashed. Copying all pending blocks to freeBlock channel for cleanup.
+	// We can clean up from uploadChannel also, but to ensure clean up happens
+	// from one place we are copying them to freeBlock channel.
+	if timedOut {
```

Comment:
Do we want to add a test case to test the timedOut scenario?

Reply:
As mentioned in the comment, its a crash scenario which i am able to replicate in the unit test. So not able to test case.

Reply:
We can add this simulated test case but the timeout is too large here:


```
func (t *UploadHandlerTest) TestDestroyWithTimeout() {
	// Simulate timeout.
	t.uh.wg.Add(1)
	b, err := t.blockPool.Get()
	require.NoError(t.T(), err)
	t.uh.uploadCh <- b

	t.uh.Destroy()

	assertAllBlocksProcessed(t.T(), t.uh)
	assert.Equal(t.T(), 1, len(t.uh.freeBlocksCh))
	assert.Equal(t.T(), 0, len(t.uh.uploadCh))
}
```

Reply:
removed the timeout flow.